python/cuml/cuml/cluster/hdbscan/hdbscan.pyx

```diff
@@ -177,156 +54,417 @@ def import_hdbscan():
+                    <size_t>temp_buffer.data(),
+                    (self.n_clusters,),
+                    np.int32,
+                    self
+                ).copy()
+            )
+        else:
+            self.inverse_label_map = CumlArray.empty((0,), dtype=np.int32)
 
-    Returns
-    -------
+        del temp_buffer
 
-    condensed_tree : hdbscan.plots.CondensedTree object
-    """
+        return self
+
+    cdef lib.CondensedHierarchy[int, float]* get_condensed_tree(self):
+        if self.hdbscan_output != NULL:
+            return &(self.hdbscan_output.get_condensed_tree())
+        return self.condensed_tree
+
+    @staticmethod
+    def from_dendrogram(dendrogram, min_cluster_size):
+        """Initialize internal state from a ScipPy dendrogram.
 
-    _children, _, _, _ = \
-        input_to_cuml_array(dendrogram[:, 0:2].astype('int32'), order='C',
-                            check_dtype=[np.int32],
-                            convert_to_dtype=(np.int32))
+        Parameters
+        ----------
+        dendrogram : array-like (size n_samples, 4)
+            Dendrogram in Scipy hierarchy format
 
-    _lambdas, _, _, _ = \
-        input_to_cuml_array(dendrogram[:, 2], order='C',
-                            check_dtype=[np.float32],
-                            convert_to_dtype=(np.float32))
+        min_cluster_size : int
+            Minimum number of children for a cluster to persist
+        """
+        cdef _HDBSCANState self = _HDBSCANState.__new__(_HDBSCANState)
+
+        children = input_to_cuml_array(
+            dendrogram[:, 0:2],
+            order='C',
+            check_dtype=[np.int32],
+            convert_to_dtype=np.int32,
+        )[0]
+
+        lambdas = input_to_cuml_array(
+            dendrogram[:, 2],
+            order='C',
+            check_dtype=[np.float32],
+            convert_to_dtype=np.float32,
+        )[0]
+
+        sizes = input_to_cuml_array(
+            dendrogram[:, 3],
+            order='C',
+            check_dtype=[np.int32],
+            convert_to_dtype=np.int32,
+        )[0]
+
+        cdef size_t n_leaves = dendrogram.shape[0] + 1
+
+        handle = Handle()
+        cdef handle_t *handle_ = <handle_t*> <size_t> handle.getHandle()
+
+        self.condensed_tree = new lib.CondensedHierarchy[int, float](handle_[0], n_leaves)
+        lib.build_condensed_hierarchy(
+            handle_[0],
+            <int*><uintptr_t>(children.ptr),
+            <float*><uintptr_t>(lambdas.ptr),
+            <int*><uintptr_t>(sizes.ptr),
+            <int>min_cluster_size,
+            n_leaves,
+            deref(self.condensed_tree)
+        )
+        return self
+
+    @staticmethod
+    cdef init_and_fit(
+        handle,
+        X,
+        lib.HDBSCANParams params,
+        DistanceType metric,
+        bool gen_min_span_tree,
+    ):
+        """Initialize internal state from a new `fit`"""
+        cdef _HDBSCANState self = _HDBSCANState.__new__(_HDBSCANState)
```

Comment:
What happens to the old `self` if `hdbscan.fit()` is called a second time? Is it just safely deleted?

Reply:
The `HDBSCAN` instance will drop references to it and the output arrays. If those arrays are no longer referenced anywhere else then that `_HDBSCANState` instance will also be dropped, otherwise it will be kept around until it's safe to do so.